pilot/pkg/networking/plugin/mixer/mixer.go

```diff
@@ -30,12 +34,14 @@ import (
 // Plugin is a mixer plugin.
 type Plugin struct{}
 
+// NewPlugin returns an ptr to an initialized mixer.Plugin.
+func NewPlugin() *Plugin {
```

Comment:
Please update to return the interface (plugin.Plugin), and get rid of the pointer receivers. The empty struct should be passed by value.

Reply:
ptr->struct: done, although for the record i disagree with this.
what is the reason for returning the interface here?

Reply:
btw. i assume you're changing Plugin name so i've left this alone.

Reply:
The reason is that there's no need to leak implementation details. Code should not depend on `authentication.Plugin`, code should depend on `plugin.Plugin`. It's to keep the code that depends on this less brittle, and more clearly shows the semantics of the situation (we're implementing a `plugin.Plugin`; that's what it is, and the specific implementation is not important). I'd actually go further and make (`authentication.`)`Plugin` a private type.

W.r.t. pointer vs value, there's a couple of reasons. The empty struct is a true singleton in golang and the compiler can do more to optimize calls with it than with a pointer (which it can't prove much about). Further, you should always prefer value receivers to pointer receivers because a method with a value receiver can always be invoked on both a value and a pointer of the correct type, while a pointer method can only ever be invoked on a pointer of the correct type. See the relevant section of [Effective Go](https://golang.org/doc/effective_go.html#pointers_vs_values).

Is there some reason you think this needs to be a pointer?

Reply:
One key reason that values beat pointers is that the compiler knows the a function called on a value type can't mutate that type for the caller, so it can generate better/faster code compared to the same function operating on a pointer.

Reply:
thanks for the detailed response.
the argument for returning a concrete type vs. interface is it allows the caller to decide which they want. for example in unit tests inside the package you may want to test private methods - that's not possible unless you have the concrete type. i think leaking should not be a big concern: that is controlled through public vs. private methods. if the implementer really wants a public method outside the interface spec let it be so (although i have a hard time imagining an argument to support that). it's probably unnecessary to use a heavy hand like returning interface to prevent folks from extending beyond the interface if they insist, it can be pushed back in code review. 
the reasons for ptrs vs. struct are: one, although right now the struct is empty, that may not remain so going forward, in fact i think it's quite likely to grow given that we are encouraging plugin authors to move more complexity out to plugins. that means that in the future we may end up passing sizeable structs around everywhere without even realizing when it happened. with ptr, the cost may be bigger than zero but it's very small and will always remain fixed going forward. i think that's the main reason why for any struct type one usually sees interfaces implemented for a ptr. 
the second reason is actually the contrary of your point above: i think it's better to leave the door open to let methods mutate the receiver. i'm not sure that the plugin model necessarily implies an immutable receiver - i can think of a couple of cases where this might not be so, collecting stats and locking. when in doubt i'd leave struct types as ptr implementations by default unless there's a good reason to do otherwise. 
having said that, i think this is really splitting hairs in this case which is why i'm totally ok to make these not ptrs (and interface vs. concrete if you want). your arguments are totally valid - it's a choice and as you're the guy who's responsible for pilot codebase i'm ok with whatever you decide in the end. realistically we are in full control here and the design doesn't have to be be agonized over - it's not a big deal to change these things going forward.
regarding CallbackListenerInputParams - i agree in principle with your reasoning. i think the reason to keep it as a ptr is that generally we want to pass structs by ptr everywhere by default for efficiency (i know the official word is use non-ptr as a first choice but just try doing a code review where you pass a struct by value with the same set of folks). actually that's the reason i changed the signature to return the slices in the original version of this PR rather than having a "mutable" parameter. because that convention leaves you free to pass by ptr everywhere for efficiency with the implicit assumption that all input params are always read only - all mutations are through return values or by methods on a mutable type. but that version got voted down, so here we are. i'd actually prefer to return to the version where mutations are returned rather than done in place.